rpc/src/module/wallet.rs

```diff
@@ -0,0 +1,563 @@
+    };
+    let db = Arc::new(RocksDB::open(&config, COLUMNS));
+    DefaultWalletStore::new(Arc::clone(&db)).start(Some("wallet_store"), notify);
+
+    WalletRpcImpl {
+        store: DefaultWalletStore::new(db),
+    }
+}
+
+impl<WS: WalletStore + 'static> WalletRpc for WalletRpcImpl<WS> {
+    fn get_live_cells_by_lock_hash(
+        &self,
+        lock_hash: H256,
+        page: usize,
+        per_page: usize,
+    ) -> Result<Vec<LiveCellWithOutPoint>> {
+        let per_page = per_page.min(50);
+        Ok(self
+            .store
+            .get_live_cells(&lock_hash, page.saturating_mul(per_page), per_page))
+    }
+
+    fn get_transactions_by_lock_hash(
+        &self,
+        lock_hash: H256,
+        page: usize,
+        per_page: usize,
+    ) -> Result<Vec<CellTransaction>> {
+        let per_page = per_page.min(50);
+        Ok(self
+            .store
+            .get_transactions(&lock_hash, page.saturating_mul(per_page), per_page))
+    }
+}
+
+pub trait WalletStore: Sync + Send {
+    fn get_live_cells(
+        &self,
+        lock_hash: &H256,
+        skip_num: usize,
+        take_num: usize,
+    ) -> Vec<LiveCellWithOutPoint>;
+
+    fn get_transactions(
+        &self,
+        lock_hash: &H256,
+        skip_num: usize,
+        take_num: usize,
+    ) -> Vec<CellTransaction>;
+}
+
+pub const WALLET_SUBSCRIBER: &str = "wallet";
+pub const COLUMNS: u32 = 3;
+
+const COLUMN_LOCK_HASH_LIVE_CELL: Col = 0;
+const COLUMN_LOCK_HASH_TRANSACTION: Col = 1;
+const COLUMN_OUT_POINT_LOCK_HASH: Col = 2;
+
+#[derive(Serialize, Deserialize, Clone)]
+struct LockHashCellOutput(H256, BlockNumber, CellOutput);
+
+fn serialize_lock_hash_key(
+    lock_hash: &H256,
+    block_number: BlockNumber,
+    tx_hash: &H256,
+    output_index: u32,
+) -> Vec<u8> {
+    let mut key = lock_hash.to_vec();
+    key.extend_from_slice(&block_number.to_le_bytes());
+    key.extend_from_slice(tx_hash.as_bytes());
+    key.extend_from_slice(&output_index.to_le_bytes());
+    key
+}
+
+fn deserialize_lock_hash_key(slice: &[u8]) -> (H256, BlockNumber, H256, u32) {
+    let lock_hash = H256::from_slice(&slice[0..32]).unwrap();
+    let block_number = BlockNumber::from_le_bytes(slice[32..40].try_into().unwrap());
+    let tx_hash = H256::from_slice(&slice[40..72]).unwrap();
+    let index = u32::from_le_bytes(slice[72..76].try_into().unwrap());
+    (lock_hash, block_number, tx_hash, index)
+}
+
+pub struct DefaultWalletStore<T> {
+    db: Arc<T>,
+}
+
+impl<T: IterableKeyValueDB + 'static> DefaultWalletStore<T> {
+    pub fn new(db: Arc<T>) -> Self {
+        DefaultWalletStore { db }
+    }
+
+    pub fn start<S: ToString>(self, thread_name: Option<S>, notify: &NotifyController) {
+        let mut thread_builder = thread::Builder::new();
+        if let Some(name) = thread_name {
+            thread_builder = thread_builder.name(name.to_string());
+        }
+
+        let new_tip_receiver = notify.subscribe_new_tip(WALLET_SUBSCRIBER);
```

Comment:
Since the DB is updated async, if the node crashes after data having been written to main DB, but before writing to the index DB, then the two DBs are out of sync. We need to store the tip in index DB, so that we know when a new tip change event comes, whether or not we should catch up the old tip first.

Reply:
aha, I totally forgot to handle this scenario, may need to add a cli command to sync wallet db with chain db.